lake/api/local.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/brimdata/zed/zio"
+	"github.com/brimdata/zed/zson"
+	"github.com/segmentio/ksuid"
+)
+
+var _ Interface = (*LocalSession)(nil)
+
+type LocalSession struct {
+	root   *lake.Root
+	engine storage.Engine
+	pools  map[ksuid.KSUID]*lake.Pool
+}
+
+func OpenLocalLake(ctx context.Context, lakePath *storage.URI) (*LocalSession, error) {
+	engine := storage.NewLocalEngine()
+	root, err := lake.Open(ctx, engine, lakePath)
+	if err != nil {
+		return nil, err
+	}
+	return &LocalSession{
+		root:   root,
+		engine: engine,
+		pools:  make(map[ksuid.KSUID]*lake.Pool),
+	}, nil
+}
+
+func CreateLocalLake(ctx context.Context, lakePath *storage.URI) (*LocalSession, error) {
+	engine := storage.NewLocalEngine()
+	root, err := lake.Create(ctx, engine, lakePath)
+	if err != nil {
+		return nil, err
+	}
+	return &LocalSession{
+		root:   root,
+		engine: engine,
+	}, nil
+}
+
+func (l *LocalSession) CreatePool(ctx context.Context, name string, layout order.Layout, thresh int64) (*lake.PoolConfig, error) {
+	if name == "" {
+		return nil, errors.New("no pool name provided")
+	}
+	pool, err := l.root.CreatePool(ctx, name, layout, thresh)
+	if err != nil {
+		return nil, err
+	}
+	return &pool.PoolConfig, nil
+}
+
+func (l *LocalSession) RenamePool(ctx context.Context, id ksuid.KSUID, name string) error {
+	if name == "" {
+		return errors.New("no pool name provided")
+	}
+	return l.root.RenamePool(ctx, id, name)
+}
+
+func (l *LocalSession) AddIndexRules(ctx context.Context, rules []index.Rule) error {
+	return l.root.AddIndexRules(ctx, rules)
+}
+
+func (l *LocalSession) DeleteIndexRules(ctx context.Context, ids []ksuid.KSUID) ([]index.Rule, error) {
+	return l.root.DeleteIndexRules(ctx, ids)
+}
+
+func (l *LocalSession) ScanIndexRules(ctx context.Context, w zio.Writer, names []string) error {
+	return l.root.ScanIndexRules(ctx, w, names)
+}
+
+func (l *LocalSession) Query(ctx context.Context, d driver.Driver, src string, srcfiles ...string) (zbuf.ScannerStats, error) {
+	query, err := compiler.ParseProc(src, srcfiles...)
+	if err != nil {
+		return zbuf.ScannerStats{}, err
+	}
+	if _, err := rlimit.RaiseOpenFilesLimit(); err != nil {
+		return zbuf.ScannerStats{}, err
+	}
+	return driver.RunWithLake(ctx, d, query, zson.NewContext(), l.root)
+}
+
+func (l *LocalSession) LookupPool(ctx context.Context, name string) (*lake.PoolConfig, error) {
+	if name == "" {
+		return nil, errors.New("no pool name provided")
+	}
+	pool := l.root.LookupPoolByName(ctx, name)
+	if pool == nil {
+		return nil, fmt.Errorf("%s: pool not found", name)
+	}
+	return pool, nil
+}
+
+func (l *LocalSession) RemovePool(ctx context.Context, id ksuid.KSUID) error {
+	return l.root.RemovePool(ctx, id)
+}
+
+func (l *LocalSession) lookupPool(ctx context.Context, id ksuid.KSUID) (*lake.Pool, error) {
+	pool, ok := l.pools[id]
+	if !ok {
+		var err error
+		pool, err = l.root.OpenPool(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		l.pools[id] = pool
+	}
+	return pool, nil
+}
```

Comment:
I put this in emulating the old design pattern of having a separate pool object, meaning you could look up the pool once and use it multiple times across other API calls.  That said, in reviewing your comment, I realized this table should just go in Lake.  Scanning the pool config log every time you want to look up a pool is silly.  And since IDs are unique there is no problem caching the state.  Whether the pool is deleted or not should not matter in the lookup method (you should be able to time travel and see the config state of a deleted pool).  The ListPools on the other hand, will give the list of pools that exist at any point in time (modulo the fundamental race between a list and delete operation).  So, net net, I think we should leave this here in this PR and move it soon, see issue #2938

Reply:
Hmm, or maybe this should just go.  This raises the fundamental consistency issue of building config state on the cloud storage model.

Reply:
Since everything is cacheable except the "HEAD" file in the journal, maybe the cache consistency check should be tied to reading HEAD.

Reply:
Ok, I closed issue #2938, opened issue #2939, and deleted the ID-to-pool cache from this PR.